---
title: "Broad TALL project "
author: "[Guillaume Charbonnier](https://guillaumecharbonnier.github.io/)"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  bookdown::gitbook:
    split_by: rmd
    toc_depth: 5
    download: ["epub", "mobi"]
    sharing:
      facebook: false
      twitter: false
---

```{r setKnitrChunkDefaultOptions, include=FALSE}
eval_after <- c(
  "fig.cap",
  "fig.height",
  "fig.width",
  "out.height",
  "out.width"
)

knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  fig.align = "center",
  eval.after = eval_after,
  cache = T
)
```


```{r loadLibraries, cache=F}
loadLibrary <- function(package) {
  if (!require(basename(package), character.only = TRUE)) BiocManager::install(package, update = FALSE)
  library(basename(package), character.only = TRUE)
}

packages <- c(
  "data.table",
  "yaml",
  "ggpubr",
  "ggrepel",
  "pheatmap",
  "limma",
  "yaml",
  "styler",
  "GenomicRanges",
  "RColorBrewer",
  "org.Hs.eg.db",
  #' BRGenomics', pausing index outside Rmd because conflit with affy R v3 VS R v4
  #' pdftools',
  #' openxlsx',
  #' biomaRt',
  #' clusterProfiler',
  #' ReactomePA',
  #' GOplot',
  #' plotly',
  #' factoextra',
  "edgeR",
  #' ggcorrplot',
  #' ggrepel',
  #' GGally',
  #' scales',
  #' UpSetR',
  #' reticulate',
  "affy",
  "affxparser",
  "barley1cdf",
  "DT",
  "R.utils", # load bz2 with fread
  "knitr"
)

invisible(lapply(packages, loadLibrary))
```

```{r loadFunctions, cache=F}
# Dependencies generated by Snakemake are defined with a path from the Metaworkflow directory using "smi" variable
# Bookdown needs to have these file available from the compilation directory and the book directory.
# This function link file from Metaworkflow directory to these two directories.
linkSmiToRmdAndBookDir <- function(smi) {
  rmd_to_mw_rel_path <- "../../.."
  smi_from_rmd <- file.path(rmd_to_mw_rel_path, smi)
  book_smi_from_rmd <- file.path(rmd_to_mw_rel_path, "out/bookdown/broad_tall", smi)
  dir.create(dirname(smi), recursive = TRUE, showWarnings = FALSE)
  dir.create(dirname(book_smi_from_rmd), recursive = TRUE, showWarnings = FALSE)

  # if (!file.exists(smi) | tools::md5sum(smi_from_rmd) != tools::md5sum(smi)) {
  if (!file.exists(smi) | file.info(smi_from_rmd)$mtime != file.info(smi)$mtime) {
    message("Copying this file: ", smi)
    # We copy date to use mtime instead of md5sum to check if file was updated
    file.copy(smi_from_rmd, smi, overwrite = TRUE, copy.date = TRUE)
    file.copy(smi_from_rmd, book_smi_from_rmd, overwrite = TRUE, copy.date = TRUE)
  }
  return(invisible(NULL))
}
# unlink("out", recursive=TRUE)
# unlink("../../../out/bookdown/plbio/out", recursive=TRUE)

# bookdown_out_dir <- file.path(mw_dir, 'out/bookdown/plbio')
# smi_out
# smi_bookdown <- file.path(bookdown_out_dir, smi)
# file.copy(smi, smi_bookdown)

# http://rstudio-pubs-static.s3.amazonaws.com/1204_67621a69f1dc465f81de9716ec063742.html
# multi.tests <- function(fun = t.test, df, vars, group.var, ...) {
# 	sapply(simplify = FALSE,                                    # sapply(simplify=T) better, elements named
# 		   vars,                                                # loop on vector of outcome variable names
# 		   function(var) {
# 			   formula <- as.formula(paste(var, "~", group.var))# create a formula with outcome and grouping var.
# 			   fun(data = df, formula, ...)                     # perform test with a given fun, default t.test
# 		   }
# 	)
# }

# Retrieve the relative path used by Bookdown as output folder
output_dir <- read_yaml("_bookdown.yml")$output_dir
rmd_to_mw_rel_path <- sub("/[^\\.].*$", "", output_dir)
outdir <- gsub("\\.\\./", "", output_dir)

source(file.path(
  rmd_to_mw_rel_path,
  "../mw-lib/src/r/functions/produceFormattedDataTable.R"
))

source("templates/applyTemplates.R")



# This function take a data object (either a matrix or a data frame)
# and produce a datatable with my favourite default settings for display.
set1Datatable <- function(data) {
  result_table <- datatable(data,
    extensions = "Buttons",
    options = list(
      dom = "Bfrtip",
      buttons = c("csv", "excel"),
      scrollX = TRUE,
      scrollY = TRUE
    ),
    class = "display nowrap",
    escape = FALSE,
    rownames = FALSE
  )
  return(result_table)
}
```

```{r loadCommonVars}
symnumargs <- list(cutpoints = c(0, 1e-04, 0.001, 0.01, 0.05, 1), symbols = c("∗∗∗∗", "∗∗∗", "∗∗", "∗", "ns"))
# symnumargs <- list(cutpoints = c(0,  1e-10,  1e-09,  1e-08,  1e-07,  1e-06,  1e-05,  1e-04, 0.001, 0.01, 0.05, 1), symbols = c("∗∗∗∗", "∗∗∗", "∗∗", "∗", "ns"))
```


```{r definePlotViolinBoxplot}
# This palette is used to fit the one from DeepTools plotHeatmap

deeptools_4_colors_palette <- c(
  "#1A1A8C",
  "#1A8CFF",
  "#8AFF87",
  "#FF9E1A"
)

plotViolinBoxplot <- function(data,
                              output_dir_from_rmd = output_dir,
                              pdf_dir_from_bookdown = file.path(
                                "pdf",
                                opts_current$get("label")
                              ),
                              y_lim = NULL,
                              facet_type = c("grid", "wrap"),
                              color_palette = deeptools_4_colors_palette,
                              comparisons_method = c("t.test", "wilcox", "none"),
                              comparisons_alternative = c("two.sided", "less", "greater"),
                              comparisons_pairs = comparisons_list,
                              outlier_alpha = 0.4) {
  comparisons_method <- match.arg(comparisons_method)
  comparisons_alternative <- match.arg(comparisons_alternative)
  facet_type <- match.arg(facet_type)

  if (is.null(color_palette)) {
    p <- ggplot(data, aes(x = peak_type, y = log2_ratio))
  } else {
    p <- ggplot(data, aes(x = peak_type, y = log2_ratio, color = peak_type))
    p <- p + scale_color_manual(values = color_palette)
  }

  if (facet_type == "grid") {
    p <- p + facet_grid(~sample)
  } else {
    p <- p + facet_wrap(
      ~sample,
      nrow = 1,
      ncol = length(unique(data$sample)),
      scales = "free_y"
    )
  }
  p <- p + geom_violin()
  p <- p + geom_boxplot(
    width = 0.3,
    outlier.alpha = outlier_alpha
  )
  if (outlier_alpha == 0) {
    fig_cap_outlier <- "Outliers are hidden. "
  } else {
    fig_cap_outlier <- ""
  }

  p <- p + stat_summary(
    fun = mean,
    colour = "darkred",
    geom = "point",
    shape = 18,
    size = 3,
    show.legend = FALSE
  )

  if (comparisons_method %in% c("t.test", "wilcox")) {
    p <- p + stat_compare_means(
      label = "p.signif",
      label.x = 1.5,
      method = comparisons_method,
      method.args = list(alternative = comparisons_alternative),
      comparisons = comparisons_pairs
    )
    if (comparisons_method == "t.test") {
      readable_test <- "T-test"
    }
    if (comparisons_method == "wilcox") {
      readable_test <- "Wilcoxon test"
    }

    fig_cap_comparisons <- paste0(
      "Comparison using ",
      readable_test,
      " with ",
      comparisons_alternative,
      " alternative. "
    )
  } else if (comparisons_method == "none") {
    fig_cap_comparisons <- ""
  }

  p <- p + theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
  if (is.null(y_lim)) {
    fig_cap_zoom <- ""
  } else {
    p <- p + coord_cartesian(ylim = y_lim)
    fig_cap_zoom <- paste0(
      "Zoom on the ",
      y_lim[1],
      ":",
      y_lim[2],
      " y-scale. "
    )
  }
  p <- p + theme_bw()
  print(p)

  pdf_dir <- file.path(
    output_dir_from_rmd,
    pdf_dir_from_bookdown
  )
  dir.create(pdf_dir, recursive = TRUE, showWarnings = FALSE)

  plot_height <- 4
  plot_width <- 8

  for (plot_height in 4:7) {
    for (plot_width in 7:10) {
      plot_filename <- paste0(
        pdf_dir,
        "/height",
        plot_height,
        "_width",
        plot_width,
        ".pdf"
      )

      ggsave(
        filename = plot_filename,
        plot = p,
        height = plot_height,
        width = plot_width
      )
    }
  }

  fig_cap <- paste0(
    fig_cap_outlier,
    fig_cap_zoom,
    fig_cap_comparisons,
    "Alternative dimensions available [here](",
    pdf_dir_from_bookdown,
    ")."
  )
  return(fig_cap)
}
```



# Abstract
Analyses done for the Broad TALL project.

# Methods

## Functional enrichment analysis (paragraph for the article)
Functional enrichment analyses were produced using a custom pipeline to automate multi-sample queries to GREAT web-service. In summary, broad and sharp peaks were queried with rGREAT R package against GO Biological Process. Genomic association rules with genes were done using default GREAT "Basal plus extension". Significant terms were filtered as those having Binomial Fold Enrichment higher than 2 and both Binomial and Hypergeometric tests Benjamini-Hochberg adjusted p-values lower than 0.05. GOSemSim R package was used to compute Wang similarity distance between all terms. Terms with a similarity distance higher than 0.8 were grouped. Only the best term according to Binomial Fold Enrichment for each group and sample was kept. Terms were further filtered for heatmap display to keep only the best 5 terms for each sample according to the binomial p-value. Color scale on heatmaps displays Binomial Benjammini-Hochberg adjusted (BH) P-values.

## Spikein analysis
Analyse 1 :
Comparer la taille de peaks de tous les échantillons (thymocytes, T-ALL & cell lines) -> Box plot

Analyse 2
Avec ChIp-seq Jurkat et Loucy alignés sur l’homme :
Panel 1: average profiles of H3K4me3 at broad peaks in Loucy and Jurkat
Panel 2: average profile of H3K4me3 in Loucy and Jurkat at common broad peaks.

Panel Panel 3-4: Meme analyse que panel 1 et 2 mais avec l’alignement sur droso. Dans ce cas prendre tous les peaks H3K4me3 trouvés (sharp et broad)

# Results
## Shiny app for functional enrichment
Shiny app is available [here](https://gcbio.info/greatr_H3K4me3_broad/). This analysis is flawed in two ways. First, it was done using original peaks from Mohammed which were duplicated for some of them. Second, it is a draft version of greatr which always keep the topN for each samples and do not keep a significant value for a sample which is outside its topN.

## Some cell lines exhibits broader H3K4me3 peaks in TSS of protein coding genes
This analysis is first done using the incomplete set of broad peaks selected from inflection point (Figure \@ref(fig:plotViolinBoxplotNMBroad)), then from MACS2 (Figure \@ref(fig:plotViolinBoxplotProteinCodingTSSBroadPeaks))

```{r loadNMBroadPeaks}
nm_broad_peaks <- list()

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/CUTLL1_peaks_broadPeak_Broad.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$CUTLL1 <- fread(smi)
nm_broad_peaks$CUTLL1[, V5 := NULL]

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_ALLT10C_H3K4me3_S01PW6H1.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT820 <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_ALLT12C_H3K4me3_S01PX4H1.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT760 <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_ALLT5C_H3K4me3_S01PSEH1.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT864 <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_ALLT6C_H3K4me3_S01PTCH1.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT885 <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_ALLT9C_H3K4me3_S01PV8H1.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT845 <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_CD4_TH91_H3K4me3_S010R5H1.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$CD4 <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_CD8_TH91_H3K4me3_S010S3H1.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$CD8 <- fread(smi)

# Not this one because the file contains only NA. The good one is the first in the list above
# smi <- 'out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_CUTLL1_H3K4me3.bed'
# linkSmiToRmdAndBookDir(smi)
# nm_broad_peaks$ <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_DND41.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$DND41 <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_EC_TH91_H3K4me3_S010NDH1.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$EC <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_Jurkat.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$Jurkat <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_LC_TH91_H3K4me3_S010Q7H1.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$LC <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_Loucy_H3K4me3.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$Loucy <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_RPMI-8402_H3K4me3.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$"RPMI-8402" <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_Sil-ALL_H3K4me3.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$"Sil-ALL" <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_T3C_H3K4me3.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT427 <- fread(smi)

smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad/NM_Broad_TALL11_H3K4me3_9886.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT802 <- fread(smi)
```

```{r plotViolinBoxplotNMBroad, fig.cap='Length distribution of broad (inflection point) H3K4me3 peaks falling into protein coding TSS.'}
d <- reshape2::melt(nm_broad_peaks, id.vars = c("V1", "V2", "V3", "V4"))
setnames(d, "L1", "Sample")
d <- data.table(unique(d))
d[, peak_length := V3 - V2]
d[, c("V1", "V2", "V3", "V4") := NULL]

cell_lines <- c("Jurkat", "RPMI-8402", "CUTLL1", "Sil-ALL", "Loucy", "DND41")
talls <- c("UPNT427", "UPNT760", "UPNT802", "UPNT820", "UPNT845", "UPNT864", "UPNT885")
thymocytes <- c("EC", "LC", "CD4", "CD8")

d[, cell_type := ifelse(Sample %in% cell_lines,
  "Cell line",
  ifelse(Sample %in% talls,
    "TALL",
    ifelse(Sample %in% thymocytes,
      "Thymocyte",
      "Other"
    )
  )
)]

d$Sample <- factor(d$Sample, levels = c(thymocytes, talls, cell_lines))

p <- ggplot(d, aes(x = Sample, y = log2(peak_length), color = cell_type))
p <- p + theme_minimal()
p <- p + geom_violin(scale = "width", adjust = 0.5, fill = "#A4A4A4", color = "#2E3436")
p <- p + geom_boxplot(width = 0.4, outlier.shape = NA, notch = TRUE)
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
# p <- p + geom_dotplot(binaxis='y', aes(colour=cellgroup, fill="#FFFFFF"), stackdir='center', dotsize=2)
# p <- p + stat_compare_means(comparisons = list(c('HDneg', 'HDpos')), method = "wilcox", alternative = "less", symnum.args = symnumargs)
p
```

```{r loadProteinCodingTSSBroadPeaks}
nm_broad_peaks <- list()

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/CEM_peaks.broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$"CCRF-CEM" <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/ALLT10C_H3K4me3_S01PW6H1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT820 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/ALLT12C_H3K4me3_S01PX4H1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT760 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/ALLT5C_H3K4me3_S01PSEH1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT864 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/ALLT6C_H3K4me3_S01PTCH1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT885 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/ALLT9C_H3K4me3_S01PV8H1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT845 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/CD4_TH91_H3K4me3_S010R5H1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$CD4 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/CD8_TH91_H3K4me3_S010S3H1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$CD8 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/CUTLL1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$CUTLL1 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/DND41_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$DND41 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/EC_TH91_H3K4me3_S010NDH1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$EC <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/Jurkat_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$Jurkat <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/LC_TH91_H3K4me3_S010Q7H1_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$LC <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/Loucy_H3K4me3_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$Loucy <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/RPMI-8402_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$"RPMI-8402" <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/Sil-ALL_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$"Sil-ALL" <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/T3C_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT427 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/T11_H3K4me3_9886_peaks_broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT802 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/T15_k4me3_peaks.broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT790 <- fread(smi)

smi <- "out/bedtools/intersect_-u_-b_bed-hg19-ensembl-r87-tss-protein-coding/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/Broad_all_genes/T16_k4me3_peaks.broadPeak.bed"
linkSmiToRmdAndBookDir(smi)
nm_broad_peaks$UPNT764 <- fread(smi)
```

```{r plotViolinBoxplotProteinCodingTSSBroadPeaks, fig.cap='Length distribution of broad (MACS2) H3K4me3 peaks falling into protein coding TSS.'}
d <- reshape2::melt(nm_broad_peaks, id.vars = names(nm_broad_peaks[[1]]))
setnames(d, "L1", "Sample")
d <- data.table(unique(d))
d[, peak_length := V3 - V2]
d[, c("V1", "V2", "V3", "V4") := NULL]

# cell_lines <- c("Jurkat","RPMI-8402","CUTLL1","Sil-ALL","Loucy","DND41")
cell_lines <- c("DND41", "Jurkat", "Loucy", "CUTLL1", "RPMI-8402", "CCRF-CEM", "Sil-ALL")

talls <- c("UPNT427", "UPNT864", "UPNT885", "UPNT845", "UPNT820", "UPNT802", "UPNT760", "UPNT790", "UPNT764")
thymocytes <- c("EC", "LC", "CD4", "CD8")

d[, cell_type := ifelse(Sample %in% cell_lines,
  "Cell line",
  ifelse(Sample %in% talls,
    "TALL",
    ifelse(Sample %in% thymocytes,
      "Thymocyte",
      "Other"
    )
  )
)]

d$Sample <- factor(d$Sample, levels = c(thymocytes, talls, cell_lines))

p <- ggplot(d, aes(x = Sample, y = log2(peak_length), color = cell_type))
p <- p + theme_minimal()
p <- p + geom_violin(scale = "width", adjust = 0.5, fill = "#A4A4A4", color = "#2E3436")
p <- p + geom_boxplot(width = 0.4, outlier.shape = NA, notch = TRUE)
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
# p <- p + geom_dotplot(binaxis='y', aes(colour=cellgroup, fill="#FFFFFF"), stackdir='center', dotsize=2)
p
```

```{r plotBoxplotProteinCodingTSSBroadPeaks, fig.cap='Length distribution of broad (MACS2) H3K4me3 peaks falling into protein coding TSS.'}
p <- ggplot(d, aes(x = Sample, y = peak_length, color = cell_type))
p <- p + theme_minimal()
p <- p + geom_boxplot(width = 0.4, outlier.shape = NA, notch = TRUE)
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

# compute lower and upper whiskers
# Inspired from https://stackoverflow.com/questions/5677885/ignore-outliers-in-ggplot2-boxplot
ylim_no_outlier <- c(
  boxplot.stats(d[Sample == "Jurkat"]$peak_length)$stats[1],
  boxplot.stats(d[Sample == "Loucy"]$peak_length)$stats[5]
)
# scale y limits based on ylim1
p <- p + coord_cartesian(ylim = ylim_no_outlier)
p
```

```{r plotViolinBoxplotProteinCodingTSSBroadPeaksWithWilcoxonTest, fig.cap='Length distribution of broad (MACS2) H3K4me3 peaks falling into protein coding TSS. Asterisks show significance of Wilcoxon test with "both" alternative. ✱ : p < 0.05, ✱✱ : p < 0.01, ✱✱✱: p < 0.001, ✱✱✱✱ : p < 0.0001. Related stats in the table below. 0 as P-value actually means it is below 1e-313.', fig.height=10}
pair_comparisons <- list(
  c("CD8", "UPNT764"),
  c("CD8", "DND41"),
  c("CD8", "Jurkat"),
  c("CD8", "Loucy"),
  c("CD8", "CUTLL1"),
  c("CD8", "RPMI-8402"),
  c("CD8", "CCRF-CEM"),
  c("CD8", "Sil-ALL"),
  c("CD8", "CD4"),
  c("CD8", "LC"),
  c("CD8", "EC")
)
p <- p + stat_compare_means(comparisons = pair_comparisons, method = "wilcox", alternative = "both", symnum.args = symnumargs)
p
```

```{r stats}
stats <- list()
stats$Thymocyte <- c(quantile(d[cell_type == "Thymocyte"]$peak_length),
  mean = mean(d[cell_type == "Thymocyte"]$peak_length),
  pval_wilcox_VS_All = wilcox.test(
    x = d$peak_length,
    y = d[cell_type == "Thymocyte"]$peak_length
  )$p.value,
  pval_wilcox_VS_Thymocyte = wilcox.test(
    x = d[cell_type == "Thymocyte"]$peak_length,
    y = d[cell_type == "Thymocyte"]$peak_length
  )$p.value,
  pval_wilcox_VS_Loucy = wilcox.test(
    x = d[Sample == "Loucy"]$peak_length,
    y = d[cell_type == "Thymocyte"]$peak_length
  )$p.value,
  pval_wilcox_VS_DND41 = wilcox.test(
    x = d[Sample == "DND41"]$peak_length,
    y = d[cell_type == "Thymocyte"]$peak_length
  )$p.value
)

stats$TALL <- c(quantile(d[cell_type == "TALL"]$peak_length),
  mean = mean(d[cell_type == "TALL"]$peak_length),
  pval_wilcox_VS_All = wilcox.test(
    x = d$peak_length,
    y = d[cell_type == "TALL"]$peak_length
  )$p.value,
  pval_wilcox_VS_Thymocyte = wilcox.test(
    x = d[cell_type == "Thymocyte"]$peak_length,
    y = d[cell_type == "TALL"]$peak_length
  )$p.value,
  pval_wilcox_VS_Loucy = wilcox.test(
    x = d[Sample == "Loucy"]$peak_length,
    y = d[cell_type == "TALL"]$peak_length
  )$p.value,
  pval_wilcox_VS_DND41 = wilcox.test(
    x = d[Sample == "DND41"]$peak_length,
    y = d[cell_type == "TALL"]$peak_length
  )$p.value
)

stats$Cell_line <- c(quantile(d[cell_type == "Cell line"]$peak_length),
  mean = mean(d[cell_type == "Cell line"]$peak_length),
  pval_wilcox_VS_All = wilcox.test(
    x = d$peak_length,
    y = d[cell_type == "Cell line"]$peak_length
  )$p.value,
  pval_wilcox_VS_Thymocyte = wilcox.test(
    x = d[cell_type == "Thymocyte"]$peak_length,
    y = d[cell_type == "Cell line"]$peak_length
  )$p.value,
  pval_wilcox_VS_Loucy = wilcox.test(
    x = d[Sample == "Loucy"]$peak_length,
    y = d[cell_type == "Cell line"]$peak_length
  )$p.value,
  pval_wilcox_VS_DND41 = wilcox.test(
    x = d[Sample == "DND41"]$peak_length,
    y = d[cell_type == "Cell line"]$peak_length
  )$p.value
)

for (sample in unique(d$Sample)) {
  stats[[sample]] <- c(quantile(d[Sample == sample]$peak_length),
    mean = mean(d[Sample == sample]$peak_length),
    pval_wilcox_VS_All = wilcox.test(
      x = d$peak_length,
      y = d[Sample == sample]$peak_length
    )$p.value,
    pval_wilcox_VS_Thymocyte = wilcox.test(
      x = d[cell_type == "Thymocyte"]$peak_length,
      y = d[Sample == sample]$peak_length
    )$p.value,
    pval_wilcox_VS_Loucy = wilcox.test(
      x = d[Sample == "Loucy"]$peak_length,
      y = d[Sample == sample]$peak_length
    )$p.value,
    pval_wilcox_VS_DND41 = wilcox.test(
      x = d[Sample == "DND41"]$peak_length,
      y = d[Sample == sample]$peak_length
    )$p.value
  )
}
stats <- as.data.frame(stats)
tstats <- t(stats)
out_table <- as.data.frame(data.table(condition = rownames(tstats), tstats))
out_table$mean <- round(out_table$mean)
set1Datatable(out_table)
```

```{r, fig.cap='Length distribution of inflexion-filtered broad H3K4me3 peaks falling into protein coding TSS. Related stats in the table below. 0 as P-value actually means it is below 1e-313.'}
smi <- "out/ln/updir/mw-tall/inp/Peaks_hg19_Mohammed/table_S2.tsv"
linkSmiToRmdAndBookDir(smi)
d_table_S2 <- fread(smi)

d <- merge(d, d_table_S2, by = "Sample")
d <- d[peak_length > Threshold_BH4D_bp]
d <- d[, c("Sample", "peak_length", "cell_type")]

d$Sample <- factor(d$Sample, levels = c(thymocytes, talls, cell_lines))

p <- ggplot(d, aes(x = Sample, y = log2(peak_length), color = cell_type))
p <- p + theme_minimal()
p <- p + geom_violin(scale = "width", adjust = 0.5, fill = "#A4A4A4", color = "#2E3436")
p <- p + geom_boxplot(width = 0.4, outlier.shape = NA, notch = TRUE)
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
# p <- p + geom_dotplot(binaxis='y', aes(colour=cellgroup, fill="#FFFFFF"), stackdir='center', dotsize=2)
p
```

```{r plotBoxplotProteinCodingTSSBroadPeaksInflexionPoint, fig.cap='Length distribution of inflexion-filtered broad H3K4me3 peaks falling into protein coding TSS. Related stats in the table below. 0 as P-value actually means it is below 1e-313.'}
p <- ggplot(d, aes(x = Sample, y = peak_length, color = cell_type))
p <- p + theme_minimal()
p <- p + geom_boxplot(width = 0.4, outlier.shape = NA, notch = TRUE)
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

# compute lower and upper whiskers
# Inspired from https://stackoverflow.com/questions/5677885/ignore-outliers-in-ggplot2-boxplot
ylim_no_outlier <- c(
  boxplot.stats(d[Sample == "UPNT764"]$peak_length)$stats[1],
  boxplot.stats(d[Sample == "DND41"]$peak_length)$stats[5]
)
# scale y limits based on ylim1
p <- p + coord_cartesian(ylim = ylim_no_outlier)
p
```

```{r}
stats <- list()
stats$Thymocyte <- c(quantile(d[cell_type == "Thymocyte"]$peak_length),
  mean = mean(d[cell_type == "Thymocyte"]$peak_length),
  pval_wilcox_VS_All = wilcox.test(
    x = d$peak_length,
    y = d[cell_type == "Thymocyte"]$peak_length
  )$p.value,
  pval_wilcox_VS_Thymocyte = wilcox.test(
    x = d[cell_type == "Thymocyte"]$peak_length,
    y = d[cell_type == "Thymocyte"]$peak_length
  )$p.value,
  pval_wilcox_VS_Loucy = wilcox.test(
    x = d[Sample == "Loucy"]$peak_length,
    y = d[cell_type == "Thymocyte"]$peak_length
  )$p.value,
  pval_wilcox_VS_DND41 = wilcox.test(
    x = d[Sample == "DND41"]$peak_length,
    y = d[cell_type == "Thymocyte"]$peak_length
  )$p.value
)

stats$TALL <- c(quantile(d[cell_type == "TALL"]$peak_length),
  mean = mean(d[cell_type == "TALL"]$peak_length),
  pval_wilcox_VS_All = wilcox.test(
    x = d$peak_length,
    y = d[cell_type == "TALL"]$peak_length
  )$p.value,
  pval_wilcox_VS_Thymocyte = wilcox.test(
    x = d[cell_type == "Thymocyte"]$peak_length,
    y = d[cell_type == "TALL"]$peak_length
  )$p.value,
  pval_wilcox_VS_Loucy = wilcox.test(
    x = d[Sample == "Loucy"]$peak_length,
    y = d[cell_type == "TALL"]$peak_length
  )$p.value,
  pval_wilcox_VS_DND41 = wilcox.test(
    x = d[Sample == "DND41"]$peak_length,
    y = d[cell_type == "TALL"]$peak_length
  )$p.value
)

stats$Cell_line <- c(quantile(d[cell_type == "Cell line"]$peak_length),
  mean = mean(d[cell_type == "Cell line"]$peak_length),
  pval_wilcox_VS_All = wilcox.test(
    x = d$peak_length,
    y = d[cell_type == "Cell line"]$peak_length
  )$p.value,
  pval_wilcox_VS_Thymocyte = wilcox.test(
    x = d[cell_type == "Thymocyte"]$peak_length,
    y = d[cell_type == "Cell line"]$peak_length
  )$p.value,
  pval_wilcox_VS_Loucy = wilcox.test(
    x = d[Sample == "Loucy"]$peak_length,
    y = d[cell_type == "Cell line"]$peak_length
  )$p.value,
  pval_wilcox_VS_DND41 = wilcox.test(
    x = d[Sample == "DND41"]$peak_length,
    y = d[cell_type == "Cell line"]$peak_length
  )$p.value
)

for (sample in unique(d$Sample)) {
  stats[[sample]] <- c(quantile(d[Sample == sample]$peak_length),
    mean = mean(d[Sample == sample]$peak_length),
    pval_wilcox_VS_All = wilcox.test(
      x = d$peak_length,
      y = d[Sample == sample]$peak_length
    )$p.value,
    pval_wilcox_VS_Thymocyte = wilcox.test(
      x = d[cell_type == "Thymocyte"]$peak_length,
      y = d[Sample == sample]$peak_length
    )$p.value,
    pval_wilcox_VS_Loucy = wilcox.test(
      x = d[Sample == "Loucy"]$peak_length,
      y = d[Sample == sample]$peak_length
    )$p.value,
    pval_wilcox_VS_DND41 = wilcox.test(
      x = d[Sample == "DND41"]$peak_length,
      y = d[Sample == sample]$peak_length
    )$p.value
  )
}

stats <- as.data.frame(stats)
tstats <- t(stats)
out_table <- as.data.frame(data.table(condition = rownames(tstats), tstats))
out_table$mean <- round(out_table$mean)
set1Datatable(out_table)
```


## Ambiguous spike-in interpretation limits ...
Some cell lines exhibits broader H3K4me3 peaks.

slightly better ChIP for Loucy... However the spike-in itself has way more reads in Loucy which could explain the broader peaks.

```{r loadSpikeData}
spike_broad_peaks <- list()

smi <- "out/ln/alias/sst/all_samples/BDGP6/bed/broad/Spike_Droso_Loucy_H3K4me3_peaks.bed"
linkSmiToRmdAndBookDir(smi)
spike_broad_peaks$Spike_Droso_Loucy <- fread(smi)

smi <- "out/ln/alias/sst/all_samples/BDGP6/bed/broad/Spike_Droso_Jurkat_H3K4me3_peaks.bed"
linkSmiToRmdAndBookDir(smi)
spike_broad_peaks$Spike_Droso_Jurkat <- fread(smi)

smi <- "out/ln/alias/sst/all_samples/GRCh38/bed/broad/Loucy_H3K4me3_Spike_peaks.bed"
linkSmiToRmdAndBookDir(smi)
spike_broad_peaks$Loucy <- fread(smi)

smi <- "out/ln/alias/sst/all_samples/GRCh38/bed/broad/Jurkat_H3K4me3_Spike_peaks.bed"
linkSmiToRmdAndBookDir(smi)
spike_broad_peaks$Jurkat <- fread(smi)
```

```{r prepareSpikeDataForBoxplot, fig.cap='Length distribution of broad H3K4me3 peaks on overall human and drosophila genomes for Loucy and Jurkat samples with spike-in.' }
d <- data.table(reshape2::melt(spike_broad_peaks, id.vars = names(spike_broad_peaks[[1]])))
setnames(d, "L1", "Sample")
d[, peak_length := V3 - V2]
d <- d[, c("Sample", "peak_length")]

# cell_lines <- c("Jurkat","RPMI-8402","CUTTL1","Sil-ALL","Loucy","DND41")
# talls <- c("UPNT427", "UPNT760","UPNT802", "UPNT820","UPNT845", "UPNT864","UPNT885")
# thymocytes <- c("EC", "LC", "CD4", "CD8")
#
# d[ , cell_type := ifelse(L1 %in% cell_lines,
# 						 'Cell line',
# 						 ifelse(L1 %in% talls,
# 								'TALL',
# 								ifelse(L1 %in% thymocytes,
# 									   'Thymocyte',
# 									   'Other')))]
#
# d$L1 <- factor(d$L1, levels=c(thymocytes, talls, cell_lines))

p <- ggplot(d, aes(x = Sample, y = log2(peak_length)))
p <- p + theme_minimal()
p <- p + geom_violin(scale = "width", adjust = 0.5, fill = "#A4A4A4", color = "darkred")
p <- p + geom_boxplot(width = 0.4, outlier.shape = NA, notch = TRUE)
p <- p + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
# p <- p + geom_dotplot(binaxis='y', aes(colour=cellgroup, fill="#FFFFFF"), stackdir='center', dotsize=2)
# p <- p + stat_compare_means(comparisons = list(c('HDneg', 'HDpos')), method = "wilcox", alternative = "less", symnum.args = symnumargs)
p
```



## QC for spike-in

* [Facet histograms](https://drive.google.com/file/d/17IvhGZum-i2vJgr6j9-NuhZhwGoKV6mU/view?usp=sharing)
* [Droso](https://gcbio.info/reports/mw-sst/out/picard/CollectInsertSizeMetrics/samtools/sam_to_bam_bai_-q_30/bowtie2/pe_BDGP6/sickle/pe_-t_sanger_-q_20/ln/alias/sst/all_samples/fastq/)
* [Human](https://gcbio.info/reports/mw-sst/out/picard/CollectInsertSizeMetrics/samtools/sam_to_bam_bai_-q_30/bowtie2/pe_hg19/sickle/pe_-t_sanger_-q_20/ln/alias/sst/all_samples/fastq/)





<!--chapter:end:index.Rmd-->

### H3K4me3 using our ChIP-Seq with Spikein

#### On all Broad and Sharp peaks from Mohammed

```{r resultsInhibitorsOurChipSeqApplyTemplate, results='asis', cache=F}
gtftk_coverage <- list()

smi <- "out/gtftk/coverage_-x_chrominfo-hg19_bed-hg19-jurkat-sharp_bw-hg19-spikein-h3k4me3-inhibitors.txt"
linkSmiToRmdAndBookDir(smi)
gtftk_coverage$sharp <- fread(smi)

smi <- "out/gtftk/coverage_-x_chrominfo-hg19_bed-hg19-jurkat-broad_bw-hg19-spikein-h3k4me3-inhibitors.txt"
linkSmiToRmdAndBookDir(smi)
gtftk_coverage$broad <- fread(smi)
gtftk_coverage <- data.table(reshape2::melt(gtftk_coverage, id.vars = names(gtftk_coverage$sharp)))

setnames(gtftk_coverage, "L1", "peak_type")

applyTemplateOurChipSeq(gtftk_coverage,
  chunk_label_prefix = opts_current$get("label")
)
```

#### On Broad and Sharp1 subset from our RNA-Seq

```{r resultsInhibitorsOurChipSeqUsingOurRnaSharp1SubsetApplyTemplate, results='asis', cache=F}
gtftk_coverage <- list()

smi <- "out/gtftk/coverage_-x_chrominfo-hg19_bed-hg19-jurkat-sharp1-from-our-rnaseq-subsets_bw-hg19-spikein-h3k4me3-inhibitors.txt"
linkSmiToRmdAndBookDir(smi)
gtftk_coverage$sharp <- fread(smi)

smi <- "out/gtftk/coverage_-x_chrominfo-hg19_bed-hg19-jurkat-broad-from-our-rnaseq-subsets_bw-hg19-spikein-h3k4me3-inhibitors.txt"
linkSmiToRmdAndBookDir(smi)
gtftk_coverage$broad <- fread(smi)

gtftk_coverage <- data.table(reshape2::melt(gtftk_coverage,
  id.vars = names(gtftk_coverage$sharp)
))

setnames(gtftk_coverage, "L1", "peak_type")

applyTemplateOurChipSeq(gtftk_coverage,
  chunk_label_prefix = opts_current$get("label")
)
```

#### On Broad and the 3 Sharp subsets from our RNA-Seq

```{r resultsInhibitorsOurChipSeqUsingOurRna3SharpSubsetsApplyTemplate, results='asis', cache=F}
gtftk_coverage <- list()

smi <- "out/gtftk/coverage_-x_chrominfo-hg19_bed-hg19-jurkat-sharp1-from-our-rnaseq-subsets_bw-hg19-spikein-h3k4me3-inhibitors.txt"
linkSmiToRmdAndBookDir(smi)
gtftk_coverage$sharp1 <- fread(smi)

smi <- "out/gtftk/coverage_-x_chrominfo-hg19_bed-hg19-jurkat-sharp2-from-our-rnaseq-subsets_bw-hg19-spikein-h3k4me3-inhibitors.txt"
linkSmiToRmdAndBookDir(smi)
gtftk_coverage$sharp2 <- fread(smi)

smi <- "out/gtftk/coverage_-x_chrominfo-hg19_bed-hg19-jurkat-sharp3-from-our-rnaseq-subsets_bw-hg19-spikein-h3k4me3-inhibitors.txt"
linkSmiToRmdAndBookDir(smi)
gtftk_coverage$sharp3 <- fread(smi)

smi <- "out/gtftk/coverage_-x_chrominfo-hg19_bed-hg19-jurkat-broad-from-our-rnaseq-subsets_bw-hg19-spikein-h3k4me3-inhibitors.txt"
linkSmiToRmdAndBookDir(smi)
gtftk_coverage$broad <- fread(smi)

gtftk_coverage <- data.table(reshape2::melt(gtftk_coverage,
  id.vars = names(gtftk_coverage$sharp1)
))

setnames(gtftk_coverage, "L1", "peak_type")

applyTemplateOurChipSeq(gtftk_coverage,
  chunk_label_prefix = opts_current$get("label")
)
```


<!--chapter:end:3-results-5-inhibitors-2-our-chipseq.Rmd-->

